libs/cardano-ledger-core/src/Cardano/Ledger/MemoBytes.hs

```diff
@@ -60,48 +65,85 @@ import Prelude hiding (span)
 --   that were used to transmit it. Important since hashes are computed
 --   from the serialization of a type, and ToCBOR instances do not have unique
 --   serializations.
-data MemoBytes t = Memo {memotype :: !t, memobytes :: ShortByteString}
-  deriving (NoThunks) via AllowThunksIn '["memobytes"] (MemoBytes t)
-
-deriving instance NFData t => NFData (MemoBytes t)
-
-deriving instance Generic (MemoBytes t)
-
-instance (Typeable t) => ToCBOR (MemoBytes t) where
-  toCBOR (Memo _ bytes) = encodePreEncoded (fromShort bytes)
-
-instance (Typeable t, FromCBOR (Annotator t)) => FromCBOR (Annotator (MemoBytes t)) where
+data MemoBytes t era = Memo'
+  { memoType :: !(t era),
+    memoByteString :: ShortByteString,
+    memoHash :: SafeHash (Crypto era) (MemoHashIndex t)
+  }
+  deriving (NoThunks) via AllowThunksIn '["memoByteString"] (MemoBytes t era)
+
+pattern Memo :: Era era => t era -> ShortByteString -> MemoBytes t era
+pattern Memo memoType memoBytes <-
+  Memo' memoType memoBytes _
+  where
+    Memo mt mb = mkMemoBytes mt (shortToLazy mb)
+
+{-# COMPLETE Memo #-}
+
+type family MemoHashIndex (t :: Type -> Type) :: Type
+
+deriving instance NFData (t era) => NFData (MemoBytes t era)
+
+deriving instance Generic (MemoBytes t era)
+
+instance (Typeable t, Typeable era) => ToCBOR (MemoBytes t era) where
+  toCBOR (Memo' _ bytes _hash) = encodePreEncoded (fromShort bytes)
+
+instance
+  ( Typeable t,
+    Typeable era,
+    FromCBOR (Annotator (t era)),
+    Era era
+  ) =>
+  FromCBOR (Annotator (MemoBytes t era))
+  where
   fromCBOR = do
     (Annotator getT, Annotator getBytes) <- withSlice fromCBOR
-    pure (Annotator (\fullbytes -> Memo (getT fullbytes) (toShort (toStrict (getBytes fullbytes)))))
+    pure (Annotator (\fullbytes -> mkMemoBytes (getT fullbytes) (getBytes fullbytes)))
+
+instance Eq (MemoBytes t era) where (Memo' _ x _) == (Memo' _ y _) = x == y
 
-instance Eq (MemoBytes t) where (Memo _ x) == (Memo _ y) = x == y
+instance Show (t era) => Show (MemoBytes t era) where show (Memo' y _ h) = show y <> "(hash " <> show h <> ")"
```

Comment:
Make it multiline and tell the world what type of hash it is instead of just `hash`:
```suggestion
instance Show (t era) => Show (MemoBytes t era) where
  show (Memo' y _ h) = show y <> " (" <> hashAlgorithmName (Just h) <> ": " <> show h <> ")"
```